### PR TITLE
[CodeCompletion] Some refactorings surrounding solver-based completion

### PR DESCRIPTION
**Pull request by ahoppen**
Perform a couple of refactoring for solver-based completion, with the goal of sharing more code between the different completion kinds:
- Sink `TypeCheckCompletionCallbacks` from Sema to IDE
  - That’s really where it belongs, there’s no reason for it to be in Sema
- Unify logic to retrieve completion expr type for all completion callbacks
  - Previously, each completion kind had its own handling of `UnresolvedType` and some had fewer fallback mechanisms to retrieve the type, have one implementation that does everything
- Check whether surrounding context supports async in all solver-based completion kinds
  - Previously, only `ExprCompletion` checked whether the surrounding context is async, I don’t have any failing test cases for the other completion kinds but it definitely makes sense for them to share the same logic to determine whether the surrounding context is async

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: lib/IDE/CompletionLookup.cpp [hunk 1]
```diff
@@ -207,6 +161,52 @@ bool swift::ide::isCompletionDeclContextLocalContext(DeclContext *DC) {
...
+      if (auto fnType = closureType->getAs<AnyFunctionType>())
+        if (fnType->isAsync())
+          return true;
+
+    // If the closure doesn't contain any async call in the body, closure itself
+    // doesn't have 'async' type even if 'async' closure is expected.
+    //   func foo(fn: () async -> Void)
+    //   foo { <HERE> }
+    // In this case, the closure is wrapped with a 'FunctionConversionExpr'
+    // which has 'async' function type.
```
Comment (xedin) [inline lib/IDE/CompletionLookup.cpp:181]: This assume a type-checked version of the AST, is that always the case?

File: include/swift/IDE/TypeCheckCompletionCallback.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+class TypeCheckCompletionCallback {
+  bool GotCallback = false;
+
+public:
+  virtual ~TypeCheckCompletionCallback() {}
+
+  /// Called for each solution produced while type-checking an expression
+  /// that the code completion expression participates in.
+  virtual void sawSolution(const constraints::Solution &solution) {
+    GotCallback = true;
```
Comment (xedin) [inline include/swift/IDE/TypeCheckCompletionCallback.h:50]: This looks suspicious - there is no way to ensure that each override calls `TypeCheckCompletionCallback::sawSolution(solution)` internally. I see that they currently do but it only takes one to go unnoticed... I'd like to suggest `virtual` to be removed from this method, and instead be used on a `protected` `sawSolutionImpl(const Solution &)` which is going to be called from this one.

File: lib/IDE/CompletionLookup.cpp (hunk 1 above)
Comment (ahoppen) [reply lib/IDE/CompletionLookup.cpp:181]: I believe so. The call side inside `CompletionLookup` is intended for the legacy code completion, which works on a type-checked AST and https://github.com/apple/swift/pull/41885/files/03d819f442f05b58f6a2b1816b2e280292db5329#diff-3088e637c4f06d0fc6cb38370068cd6fa6fe8dcaa10ec35c9086602c59eaba17R149 works on the already type-checked part of the AST to determine async-ness after failing to determine async-ness from the solution.

File: include/swift/IDE/TypeCheckCompletionCallback.h (hunk 2 above)
Comment (ahoppen) [reply include/swift/IDE/TypeCheckCompletionCallback.h:50]: Sounds like a good idea. I’ll do that.

Comment (ahoppen) [reply include/swift/IDE/TypeCheckCompletionCallback.h:55]: Implemented in https://github.com/apple/swift/pull/41885/commits/3597652960253685386103751dad1837fe410213

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: include/swift/IDE/ArgumentCompletion.h [hunk 3]
```diff
@@ -63,13 +63,13 @@ class ArgumentTypeCheckCompletionCallback : public TypeCheckCompletionCallback {
                          SmallVectorImpl<PossibleParamInfo> &Params,
                          SmallVectorImpl<Type> &Types);
 
+  void sawSolutionImpl(const constraints::Solution &solution) override;
```
Comment (xedin) [inline include/swift/IDE/ArgumentCompletion.h:66]: Nit: mark it as pure via ` = 0;`

**Review by xedin — APPROVED** [review]

File: include/swift/IDE/ArgumentCompletion.h (hunk 3 above)
Comment (xedin) [reply include/swift/IDE/ArgumentCompletion.h:66]: Oops, wrong place :)

**Review by xedin — APPROVED** [review]

**Pull request merged**